chef/cookbooks/provisioner/libraries/find_server.rb

```diff
@@ -0,0 +1,32 @@
+#
+# Copyright 2013-2014, SUSE LINUX Products GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+class Chef
+  class Recipe
+    def provisioner_server_node
+      @provisioner_server_node ||=
+        begin
+          env = node[:provisioner][:config][:environment]
+          server = search(
+            :node,
+            "roles:provisioner-server AND provisioner_config_environment:#{env}"
+          ).first
+          Chef::Log.info("Provisioner server is #{server[:hostname]}")
+          server
+        end
+    end
```

Comment:
I'm a bit reluctant about this, because you're caching a full node. As we've seen, the cache will stay alive forever in the chef-client daemon, and I'm guessing (although not sure) that the node attributes will slowly be outdated after a while, but the cache never invalidated.

At the minimum please add the hack using ohai_time to invalidate the cache.